service-registry/registry-etcd/src/main/java/org/apache/servicecomb/registry/etcd/EtcdDiscovery.java

```diff
@@ -0,0 +1,158 @@
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.servicecomb.registry.etcd;
+
+import com.google.common.collect.Lists;
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.Client;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Watch;
+import io.etcd.jetcd.kv.GetResponse;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.WatchOption;
+import io.etcd.jetcd.watch.WatchEvent;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.servicecomb.config.BootStrapProperties;
+import org.apache.servicecomb.foundation.common.utils.JsonUtils;
+import org.apache.servicecomb.registry.api.Discovery;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.env.Environment;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+public class EtcdDiscovery implements Discovery<EtcdDiscoveryInstance> {
+
+    private Environment environment;
+
+    private String basePath;
+
+    private EtcdRegistryProperties etcdRegistryProperties;
+
+    private Client client;
+
+    private InstanceChangedListener<EtcdDiscoveryInstance> instanceChangedListener;
+
+    @Autowired
+    @SuppressWarnings("unused")
+    public void setEnvironment(Environment environment) {
+        this.environment = environment;
+    }
+
+    @Autowired
+    @SuppressWarnings("unused")
+    public void setEtcdRegistryProperties(EtcdRegistryProperties etcdRegistryProperties) {
+        this.etcdRegistryProperties = etcdRegistryProperties;
+    }
+
+    @Override
+    public String name() {
+        return EtcdConst.ETCD_REGISTRY_NAME;
+    }
+
+    @Override
+    public boolean enabled(String application, String serviceName) {
+        return environment.getProperty(String.format(EtcdConst.ETCD_DISCOVERY_ENABLED, application, serviceName),
+                boolean.class, true);
+    }
+
+    @Override
+    public List<EtcdDiscoveryInstance> findServiceInstances(String application, String serviceName) {
+
+        String prefixPath = basePath + "/" + application + "/" + serviceName;
+        Watch watchClient = client.getWatchClient();
+        watchClient.watch(ByteSequence.from(prefixPath, Charset.defaultCharset()),
+                WatchOption.builder().build(),
+                resp -> {
```

Comment:
`wacth` only need to create once, while findServiceInstances maybe invoked several times.